src/webgpu/api/operation/render_pipeline/sample_mask.spec.ts

```diff
@@ -1,16 +1,443 @@
+    sampleCount: number,
+    rasterizationMask: number,
+    sampleMask: number,
+    fragmentShaderOutputMask: number
+  ) {
+    const expected = {
+      R: 0,
+      G: 0,
+      B: 0,
+      A: 0,
+    };
+    range(4, (id: number) => {
+      const m = rasterizationMask & sampleMask & fragmentShaderOutputMask & (1 << id);
+      // WebGPU only support up to 4 samples, so samples after the first 4 should be ignored.
+      const s = (m & 0xf) === 0 ? kEmptySample : kColors[id % 4];
+      expected.R += s.R;
+      expected.G += s.G;
+      expected.B += s.B;
+      expected.A += s.A;
+    });
+
+    this.expectSinglePixelBetweenTwoValuesIn2DTexture(
+      texture,
+      format,
+      { x: 0, y: 0 },
+      {
+        exp: [
+          new Uint8Array([
+            Math.floor(expected.R / sampleCount),
+            Math.floor(expected.G / sampleCount),
+            Math.floor(expected.B / sampleCount),
+            Math.floor(expected.A / sampleCount),
+          ]),
+          new Uint8Array([
+            Math.ceil(expected.R / sampleCount),
+            Math.ceil(expected.G / sampleCount),
+            Math.ceil(expected.B / sampleCount),
+            Math.ceil(expected.A / sampleCount),
+          ]),
+        ],
+      }
+    );
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('final_output')
+  .desc(
+    `
+Tests that the final sample mask is the logical AND of all the relevant masks -- meaning that the samples
+not included in the final mask are discarded on any attachments including
+- color outputs
 TODO:
+- depth tests
+- stencil operations
+
+The test draws 0/1/1+ textured quads of which each sample in the standard 4-sample pattern results in a different color:
+- Sample 0, Texel 0, top-left: Red
+- Sample 1, Texel 1, top-left: Green
+- Sample 2, Texel 2, top-left: Blue
+- Sample 3, Texel 3, top-left: Yellow
+
+The test checks which sample is passed by looking at the final color of the 1x1 resolve target texture, to see if that matches
+what is expected given by the rasterization mask, sample mask, and fragment shader output mask.
+
 - for sampleCount = { 1, 4 } and various combinations of:
-    - rasterization mask = { 0, 1, 2, 3, 15 }
+    - rasterization mask = { 0, 1, 2, 3, 4, 5, 15 }
     - sample mask = { 0, 1, 2, 3, 15, 30 }
     - fragment shader output mask (SV_Coverage) = { 0, 1, 2, 3, 15, 30 }
-- test that final sample mask is the logical AND of all the
-  relevant masks -- meaning that the samples not included in the final mask are discarded
-  for all the { color outputs, depth tests, stencil operations } on any attachments.
 - [choosing 30 = 2 + 4 + 8 + 16 because the 5th bit should be ignored]
-`;
+`
+  )
+  .params(u =>
+    u
+      .combine('sampleCount', [1, 4] as const)
+      .beginSubcases()
+      .combine('rasterizationMask', [0, 1, 2, 3, 4, 5, 8, 15] as const)
+      .combine('sampleMask', [0, 1, 2, 3, 15, 30] as const)
+      .combine('fragmentShaderOutputMask', [0, 1, 2, 3, 15, 30] as const)
```

Comment:
nit: I know the original notes suggested this parameterization but it's probably unnecessarily big. On my machine the tests take about 500ms, which isn't terrible, but probably isn't providing much coverage. Maybe we could just do [0b0111, 0b1011, 0b1101, 0b1110]` for each parameter or something?

Reply:
I think mask = 0 is needed
mask = 0b1111 is needed.
fragmentShaderOutputMask, sampleMask = 30 is still needed to test if 5th bit is ignored.
mask = 0b0001 and 0b0010 are needed because they have different meaning for single/multisampled cases.

then we can have [0b0111, 0b1011, 0b1101, 0b1110]
which doesn't seem like we trimmed down many cases.

If the performance is a concern, as I'm still adding depth and stencil tests which will make tests running time bigger, What about putting rasterizationMask before beginSubcases. (On my machine each case now still takes 200~400ms though)

Reply:
No worries, I think we can keep as is.